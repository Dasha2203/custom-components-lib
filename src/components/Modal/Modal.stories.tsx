import { useState } from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import Modal from './Modal';
import Button from '../Button/Button';
import { within, userEvent, screen } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

const meta: Meta<typeof Modal> = {
  title: 'Components/Modal',
  component: Modal,
  tags: ['autodocs'],
  argTypes: {
    open: { control: 'boolean' },
    onClose: { action: 'closed' },
  },
};
export default meta;

type Story = StoryObj<typeof Modal>;

export const Default: Story = {
  render: (args) => {
    const [isOpen, setIsOpen] = useState(args.open);

    return (
      <div>
        <Button onClick={() => setIsOpen(true)}>Open Modal</Button>
        <Modal open={isOpen} onClose={() => setIsOpen(false)}>
          <p>This is a modal content.</p>
        </Modal>
      </div>
    );
  },
  args: {
    open: false,
  },
};

export const InteractionTest = {
  ...Default,
  play: async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);
    const openButton = await canvas.getByRole('button', {
      name: /open modal/i,
    });

    await userEvent.click(openButton);
    await expect(
      screen.getByText(/this is a modal content/i)
    ).toBeInTheDocument();

    const overlay = await screen.getByTestId('modal-overlay');
    await userEvent.click(overlay);
    await expect(
      screen.queryByText(/this is a modal content/i)
    ).not.toBeInTheDocument();

    await userEvent.click(openButton);
    await expect(
      screen.getByText(/this is a modal content/i)
    ).toBeInTheDocument();

    await userEvent.keyboard('{Escape}');
    await expect(
      screen.queryByText(/this is a modal content/i)
    ).not.toBeInTheDocument();
  },
};
