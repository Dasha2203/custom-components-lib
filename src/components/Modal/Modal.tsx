import React, { useEffect } from 'react';
import * as classes from './styles.module.scss';
import Portal from '../Portal/Portal';
import { ModalProps } from './types';

const Modal = ({ open, onClose, children }: ModalProps) => {
  const handleClose = (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    onClose();
  };

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [onClose]);

  if (!open) return null;

  return (
    <Portal>
      <div className={classes.modal} onClick={handleClose}>
        <div
          className={classes.modal__content}
          onClick={(e) => e.stopPropagation()}
          tabIndex={1}
        >
          {children}
        </div>
      </div>
    </Portal>
  );
};

export default Modal;
