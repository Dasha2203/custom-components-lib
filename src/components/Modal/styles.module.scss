.modal {
  position: fixed;
  z-index: 1300;
  inset: 0px;

  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.5);
  -webkit-tap-highlight-color: transparent;

  &__content {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    padding: 32px;
    min-width: 400px;
    width: fit-content;
    box-sizing: border-box;
    background-color: rgb(255, 255, 255);
    border: 2px solid rgb(0, 0, 0);
    box-shadow:
      rgba(0, 0, 0, 0.2) 0px 11px 15px -7px,
      rgba(0, 0, 0, 0.14) 0px 24px 38px 3px,
      rgba(0, 0, 0, 0.12) 0px 9px 46px 8px;
  }

  @media (max-width: 400px) {
    &__content {
      min-width: auto;
      width: 100%;
      inset: 0;
      transform: translate(0);
    }
  }
}
