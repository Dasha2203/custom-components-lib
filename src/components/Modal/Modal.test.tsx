import { render, screen, fireEvent } from '@testing-library/react';
import Modal from './Modal';

describe('Testing Modal component', () => {
  let handleClose: jest.Mock;

  beforeEach(() => {
    handleClose = jest.fn();
  });

  test('modal renders in the portal', () => {
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'react-portal-wrapper');
    document.body.appendChild(modalRoot);

    render(
      <Modal open={true} onClose={handleClose}>
        <p>Modal Content</p>
      </Modal>
    );

    const modalElement = screen.getByTestId('modal-overlay');
    expect(modalElement).toBeInTheDocument();
    expect(modalRoot.contains(modalElement));
    expect(screen.getByText('Modal Content')).toBeInTheDocument();

    document.body.removeChild(modalRoot);
  });

  test('closes when clicking outside the modal', () => {
    render(
      <Modal open={true} onClose={handleClose}>
        <p>Modal Content</p>
      </Modal>
    );

    expect(screen.getByText('Modal Content')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('modal-overlay'));
    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  test('closes when pressing Escape', () => {
    render(
      <Modal open={true} onClose={handleClose}>
        <p>Modal Content</p>
      </Modal>
    );

    fireEvent.keyDown(document, { key: 'Escape' });
    expect(handleClose).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshots of Modal', () => {
  let modalRoot: HTMLElement;

  beforeEach(() => {
    modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'react-portal-wrapper');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    document.body.removeChild(modalRoot);
  });

  test('Opened modal', async () => {
    render(
      <Modal open={true} onClose={jest.fn()}>
        <p>Modal content</p>
      </Modal>
    );

    const modalComponent = screen.getByTestId('modal-overlay');
    expect(modalComponent).toMatchSnapshot();
  });
});
