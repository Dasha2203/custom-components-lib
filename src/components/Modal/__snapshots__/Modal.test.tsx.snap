// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots of Modal Opened modal 1`] = `
<div
  class="modal"
  data-testid="modal-overlay"
>
  <div
    class="modal__content"
    tabindex="1"
  >
    <p>
      Modal content
    </p>
  </div>
</div>
`;
